pyuvdata/uvdata/uvdata.py

```diff
@@ -10488,109 +10483,239 @@ def frequency_average(
+                some_uneven = True
+            if keep_ragged:
+                this_final_nchan = int(np.ceil(nchans_spw[spw_ind] / n_chan_to_avg))
             else:
-                shape_tuple = (self.Nblts, 1, n_final_chan, n_chan_to_avg, self.Npols)
+                this_final_nchan = int(np.floor(nchans_spw[spw_ind] / n_chan_to_avg))
+            final_spw_chans[spw] = np.arange(
+                final_nchan, final_nchan + this_final_nchan
+            )
+            final_nchan += this_final_nchan
+
+        if some_uneven and kr_use_default:
+            warnings.warn(
+                "Some spectral windows do not divide evenly by `n_chan_to_avg` and the "
+                "keep_ragged parameter was not set. The current default is False, but "
+                "that will be changing to True in version 2.5. Set `keep_ragged` to "
+                "True or False to silence this warning.",
+                DeprecationWarning,
+            )
+        if some_uneven and keep_ragged and reset_cs:
+            warnings.warn(
+                "Ragged frequencies will result in varying channel widths, so "
+                "this object will have future array shapes after averaging. "
+                "Use keep_ragged=False to avoid this."
+            )
 
-            mask = self.flag_array.reshape(shape_tuple)
+        final_freq_array = np.zeros(final_nchan, dtype=float)
+        final_channel_width = np.zeros(final_nchan, dtype=float)
+        final_flex_spw_id_array = np.zeros(final_nchan, dtype=int)
+        if self.eq_coeffs is not None:
+            final_eq_coeffs = np.zeros((self.Nants_telescope, final_nchan), dtype=float)
 
-            if propagate_flags:
-                # if any contributors are flagged, the result should be flagged
-                if self.future_array_shapes:
-                    self.flag_array = np.any(
-                        self.flag_array.reshape(shape_tuple), axis=2
-                    )
+        if not self.metadata_only:
+            final_shape_tuple = (self.Nblts, final_nchan, self.Npols)
+            final_flag_array = np.full(final_shape_tuple, False, dtype=bool)
+            final_data_array = np.zeros(final_shape_tuple, dtype=self.data_array.dtype)
+            final_nsample_array = np.zeros(
+                final_shape_tuple, dtype=self.nsample_array.dtype
+            )
+
+        for spw_ind, spw in enumerate(self.spw_array):
+            n_final_chan_reg = int(np.floor(nchans_spw[spw_ind] / n_chan_to_avg))
+            nfreq_mod_navg = nchans_spw[spw_ind] % n_chan_to_avg
+            these_inds = spw_chans[spw]
+            this_ragged = False
+            regular_inds = these_inds
+            irregular_inds = np.array([])
+            this_final_reg_inds = final_spw_chans[spw]
+            if nfreq_mod_navg != 0:
+                # not an even number of final channels
+                regular_inds = these_inds[0 : n_final_chan_reg * n_chan_to_avg]
```

Comment:
I agree that it would be nice to allow the user to have more detailed control. We thought about something like you're suggesting, but as we thought about it seemed more complicated than just specifying where you start, e.g. what do you do with all the frequencies before the starting place? It seems to me that the most flexible thing would be to allow users to pass in an integer array that specifies which final channel to put each input channel into (and allowing the user to drop channels by specifying -1 in this array). But we decided to just do the obvious incremental improvement for now and wait to see what users request.